torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -619,6 +586,120 @@ std::vector<at::Tensor> FusionSegmentRuntime::runWithInput(
   return fusion_output_tensors;
 }
 
+const std::vector<FusionSegmentRuntime::SchedulerEntryPtr>&
+FusionSegmentRuntime::schedulers() {
+  return heuristics_->heuristics();
+}
+
+namespace {
+using HashType = FusionSegmentRuntime::HashType;
+// Use a slightly more nontrivial combine to avoid collision
+//  (from Boost)
+inline HashType combineHash(HashType a, HashType b) {
+  return a ^ (b + 0x9e3779b9 + (a << 6) + (a >> 2));
+}
+} // namespace
+
+FusionSegmentRuntime::HashType FusionSegmentRuntime::getHash(
+    SegmentHeuristics* sh) {
+  HashType h = 0;
+  for (auto& se_pt : sh->heuristics()) {
+    h = combineHash(h, SchedulerEntryHash()(*se_pt));
+  }
+  return h;
+}
+
+FusionSegmentRuntime::EntryTag::EntryTag(SegmentHeuristics* sh) {
+  heuristics_ = sh;
+  hash_ = FusionSegmentRuntime::getHash(sh);
+}
+
+bool FusionSegmentRuntime::EntryTag::operator==(
+    const FusionSegmentRuntime::EntryTag& other) const {
+  if (heuristics_->heuristics().size() !=
+      other.heuristics_->heuristics().size()) {
+    return false;
+  }
+
+  auto& heuristics = heuristics_->heuristics();
+  return std::equal(
+      heuristics.begin(),
+      heuristics.end(),
+      other.heuristics_->heuristics().begin(),
+      [](const SchedulerEntryPtr& a, const SchedulerEntryPtr& b) {
+        return a->sameAs(b.get());
+      });
+}
+
+void FusionSegmentRuntimeCache::evictId(size_t input_id) {
+  TORCH_INTERNAL_ASSERT(id_to_rt_.count(input_id) != 0);
+
+  // Evict the stored input tensor meta data
+  //  corresponding to input_id
+  id_to_rt_.at(input_id)->evictCache(input_id);
+  id_to_rt_.erase(input_id);
+}
+
+FusionSegmentRuntime* FusionSegmentRuntimeCache::getRTById(size_t input_id) {
+  if (id_to_rt_.count(input_id) == 0) {
+    return nullptr;
+  }
+  return id_to_rt_.at(input_id);
+}
+
+FusionSegmentRuntime* FusionSegmentRuntimeCache::getRTByHeuristics(
+    const at::ArrayRef<IValue>& inputs,
+    size_t input_id) {
+  auto dev_id = getCommonDeviceCUDA(inputs);
+  auto heuristics = segmented_fusion_->makeHeuristics(inputs);
+  EntryTag tag(heuristics.get());
```

Comment:
can we call it a heuristic tag instead of entry tag? Or is there other information stored here too?

Reply:
Looks like you're caching directly to concrete input shapes.

Reply:
Updated naming, thanks! The concrete input shapes are cached inside the `FusionExecutor`s contained in `FusionSegmentRuntime`. The `FusionSegmentRuntimeCache` only create entries on lists of heuristics but provides short-cuts for concrete sizes.

Reply:
Updated Thanks!